src/App.js

```diff
@@ -92,71 +91,66 @@ export default class App extends React.Component<{}, AppState> {
             drawingEnabled: true
         };
 
-        this.interpreter = new Interpreter(this.handleRunningStateChange);
+        this.commandNames = [
+            'forward1', 'forward2', 'forward3',
+            'left45', 'left90', 'left180',
+            'right45', 'right90', 'right180'
+        ];
 
-        this.interpreter.addCommandHandler(
-            'forward',
-            'moveCharacter',
-            () => {
-                this.audioManager.playSound('forward');
-                this.setState((state) => {
-                    return {
-                        characterState: state.characterState.forward(
-                            state.sceneGridCellWidth,
-                            state.drawingEnabled
-                        )
-                    };
-                });
-                return new Promise((resolve, reject) => {
-                    setTimeout(() => {
-                        resolve();
-                    }, 1750);
-                });
-            }
-        );
-
-        this.interpreter.addCommandHandler(
-            'left',
-            'moveCharacter',
-            () => {
-                this.audioManager.playSound('left');
-                this.setState((state) => {
-                    return {
-                        characterState: state.characterState.turnLeft(1)
-                    };
-                });
-                return new Promise((resolve, reject) => {
-                    setTimeout(() => {
-                        resolve();
-                    }, 1750);
-                });
-            }
-        );
+        this.interpreter = new Interpreter(this.handleRunningStateChange);
 
-        this.interpreter.addCommandHandler(
-            'right',
-            'moveCharacter',
-            () => {
-                this.audioManager.playSound('right');
-                this.setState((state) => {
-                    return {
-                        characterState: state.characterState.turnRight(1)
-                    };
-                });
-                return new Promise((resolve, reject) => {
-                    setTimeout(() => {
-                        resolve();
-                    }, 1750);
-                });
-            }
-        );
+        for(let i=0; i<this.commandNames.length; i++) {
```

Comment:
I'm not keen on this code to register the command handlers. I would prefer to follow the straightforward no-logic code that we used before. It's more verbose, but less error-prone and brittle. It's likely that we will soon 1) change the format of the commands to use parameters and 2) add new commands. When we make either of those changes we will have to make sure they work with this logic, or remove this logic. This code is baking the wrong model into the codebase (that command names are parsed to interpret their meaning). And it will make the code brittle if is persists for long.

Reply:
Thank you Simon for reviewing this PR! I think that this logic helps us to reduce duplication of registering similar commands, and it's not too complex logic -- as you know, what it does is to loop through all the available command names and register handler for the commands. I know there's some logic involved in converting degree to the new direction value, and. splitting commands' name and value followed by, and I am thinking we could refactor them out to a separate file and write test cases for them to ensure logics are working as we expected. To your point 1, when we change the format of the commands to use parameters, we can directly use those parameters to the command registration handler, and that also helps us to simplify the logic by not have to use split values. For the point 2, when we add new commands, we can simply add new commands to the commandName property and update the switch statement with new commands, or if the new commands are existing, but different values, we don't have to change the switch statement at all.

Reply:
It does reduce duplication but it introduces a model for the meaning of the command names that is brittle: that we parse the command names to understand what they do. In the future, we could automate the registration of the command handlers, in the manner similar to what you do here, but with a richer, more flexible model of command properties. But I think that it is too early to make the right model as we know that our command structure will be changing. And I don't what to invest too much time getting the model right for something that will change.

Reply:
It also makes it hard to change the time delay for different commands, without adding more logic. As we discussed the possibility of doing with Tony earlier today. In effect what you have is a single command handler that parses its command name to see what it should do. We have a lot more flexibility if command handlers are actually different.

Reply:
The `switch` statement is executed for every command invocation.

Reply:
Right, yeah it's too early for building a model that will be changed in the next release. Also, I agree with you that making individual command registration for each command make it easier for us in setting different delay values for different turning angles, I did not consider that into implementation of this model. I will go ahead and make individual handler registrations.

Reply:
Thanks Daniel. I know it's verbose but I think it's our best option in the short term.

Reply:
No problem and thanks for your comments! 👍

Reply:
An idea to reduce some of our duplication: we could introduce a utility function for making delay Promises (to `Utils.js`).

So that we could replace the following lines:

    return new Promise((resolve, reject) => {
        setTimeout(() => {
            resolve();
        }, 1750);
    });

with a single line (and still keep full flexibility):

    return makeDelayPromise(1750);